### PR TITLE
Fix wait for JSON elements appearance when response body contains not…

### PR DESCRIPTION
**Pull request by ALegchilov**
Fix wait for JSON elements appearance when response body contains not JSON

### CODE REVIEW

File: vividus-plugin-rest-api/src/test/java/org/vividus/bdd/steps/api/JsonResponseValidationStepsTests.java [hunk 1]
```diff
@@ -105,6 +106,9 @@
     private static final String JSON_PATH = "$..value";
     private static final String THE_NUMBER_OF_JSON_ELEMENTS_ASSERTION_MESSAGE =
             "The number of JSON elements by JSON path: ";
+    private static final String HTML =
+            "<html><head><meta http-equiv=\"Content-Type\" content=\"text/html;charset=utf-8\"/>";
+    private static final String PT_1_S = "PT1S";
```
Comment (ikalinin1) [inline vividus-plugin-rest-api/src/test/java/org/vividus/bdd/steps/api/JsonResponseValidationStepsTests.java]: `PT_1_S ` -> ONE_SECOND ? and `Duration.ofSeconds(1)` ?

Comment (ALegchilov) [reply vividus-plugin-rest-api/src/test/java/org/vividus/bdd/steps/api/JsonResponseValidationStepsTests.java]: > `PT_1_S ` -> ONE_SECOND ? and `Duration.ofSeconds(1)` ?
Updated

File: vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/JsonResponseValidationSteps.java [hunk 2]
```diff
@@ -233,16 +241,23 @@ public void waitForJsonFieldAppearance(String jsonPath, String resourceUrl, Dura
...
+            {
+                if (response == null || getElementsNumber(response.getResponseBodyAsString(), jsonPath) > 0)
+                {
+                    return false;
+                }
+            }
+            catch (InvalidJsonException ignored)
             {
-                return false;
+                return true;
```
Comment (valfirst) [inline vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/JsonResponseValidationSteps.java]: what will happen if the endpoint will always return response with non JSON content type? it looks like it will become an infinite loop, won't it?

Comment (valfirst) [reply vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/JsonResponseValidationSteps.java]: also there is no `sleep` after response with non JSON content type

File: vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/JsonResponseValidationSteps.java [hunk 3]
```diff
@@ -130,6 +131,13 @@ public boolean doesJsonPathElementsFromJsonMatchRule(String json, String jsonPat
         return assertJsonElementsNumber(jsonPath, actualNumber, comparisonRule, elementsNumber);
     }
 
+    private boolean doesJsonPathElementsFromJsonMatchRule(String json, String jsonPath,
```
[the remark this answers is not in the document]
Comment (ALegchilov) [reply vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/JsonResponseValidationSteps.java]: Got it.

File: vividus-plugin-rest-api/src/test/java/org/vividus/bdd/steps/api/JsonResponseValidationStepsTests.java [hunk 4]
```diff
@@ -332,6 +335,45 @@ void testWaitForJsonFieldAppearsWithPolling() throws IOException, IllegalAccessE
...
+        jsonResponseValidationSteps.waitForJsonFieldAppearance(STRING_PATH, URL, Duration.ofSeconds(1),
+                DURATION_DIVIDER);
+        verify(httpClient, times(2)).execute(argThat(base -> base instanceof HttpRequestBase),
+                argThat(context -> context instanceof HttpClientContext));
+        verify(softAssert).assertThat(eq(THE_NUMBER_OF_JSON_ELEMENTS_ASSERTION_MESSAGE + STRING_PATH), eq(1),
+                argThat(m -> "a value greater than <0>".equals(m.toString())));
+    }
+
+    @Test
+    void testSoftAssertInvalidJSONException() throws IOException, NoSuchFieldException, IllegalAccessException
```
[the remark this answers is not in the document]
Comment (ALegchilov) [reply vividus-plugin-rest-api/src/test/java/org/vividus/bdd/steps/api/JsonResponseValidationStepsTests.java]: Got you

File: vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/JsonResponseValidationSteps.java (hunk 2 above)
Comment (ALegchilov) [reply vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/JsonResponseValidationSteps.java]: No need to return something there, I will exclude the return statement.

File: vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/JsonResponseValidationSteps.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (ALegchilov) [reply vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/JsonResponseValidationSteps.java]: Is there a better way?

File: vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/JsonResponseValidationSteps.java [hunk 5]
```diff
@@ -339,7 +347,7 @@ public void performAllStepsForProvidedJsonIfFound(ComparisonRule comparisonRule,
         {
             return Optional.of(Optional.ofNullable(JsonPathUtils.getData(json, jsonPath)));
         }
-        catch (PathNotFoundException e)
+        catch (PathNotFoundException | InvalidJsonException e)
```
Comment (valfirst) [inline vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/JsonResponseValidationSteps.java]: could you please explain the reason of this change?

Comment (ikalinin1) [reply vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/JsonResponseValidationSteps.java]: If we are executing some request and after the last iteration, it is still not valid JSON we should not break the step but fail it;

File: vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/JsonResponseValidationSteps.java [hunk 6]
```diff
@@ -233,16 +234,23 @@ public void waitForJsonFieldAppearance(String jsonPath, String resourceUrl, Dura
...
+            {
+                //ignore exception
             }
             Sleeper.sleep(pollingTimeoutMillis, TimeUnit.MILLISECONDS);
             return System.currentTimeMillis() <= expectedTime;
         });
         if (httpTestContext.getResponse() != null)
         {
-            doesJsonPathElementsMatchRule(jsonPath, ComparisonRule.GREATER_THAN, 0);
+            doesJsonPathElementsFromJsonMatchRule(getActualJson(), jsonPath, ComparisonRule.GREATER_THAN, 0, true);
```
Comment (valfirst) [inline vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/JsonResponseValidationSteps.java]: is this change still needed?

Comment (ALegchilov) [reply vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/JsonResponseValidationSteps.java]: excluded

File: vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/JsonResponseValidationSteps.java [hunk 7]
```diff
@@ -233,16 +234,23 @@ public void waitForJsonFieldAppearance(String jsonPath, String resourceUrl, Dura
...
+            {
+                //ignore exception
             }
             Sleeper.sleep(pollingTimeoutMillis, TimeUnit.MILLISECONDS);
             return System.currentTimeMillis() <= expectedTime;
         });
         if (httpTestContext.getResponse() != null)
         {
-            doesJsonPathElementsMatchRule(jsonPath, ComparisonRule.GREATER_THAN, 0);
+            doesJsonPathElementsFromJsonMatchRule(getActualJson(), jsonPath, ComparisonRule.GREATER_THAN, 0);
```
Comment (valfirst) [inline vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/JsonResponseValidationSteps.java]: just `doesJsonPathElementsMatchRule(jsonPath, ComparisonRule.GREATER_THAN, 0);` ?

File: vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/JsonResponseValidationSteps.java [hunk 8]
```diff
@@ -371,13 +379,23 @@ public void performAllStepsForProvidedJsonIfFound(ComparisonRule comparisonRule,
 
     private int getElementsNumber(String json, String jsonPath)
     {
-        Optional<List<?>> elements = getElements(json, jsonPath);
```
Comment (valfirst) [inline vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/JsonResponseValidationSteps.java]: it looks like all the changes below are not required anymore

Comment (ALegchilov) [reply vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/JsonResponseValidationSteps.java]: Okay.

File: vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/JsonResponseValidationSteps.java (hunk 7 above)
Comment (ALegchilov) [reply vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/JsonResponseValidationSteps.java]: Yep
